File: benchmark_scripts/nbody-par/jl/nbody.jl
using Random
using Base.Threads

DELTA_TIME=1.0
CONST_G=1.0

mutable struct Particle
    mass::Float64
    posx::Float64
    posy::Float64
    velx::Float64
    vely::Float64
    accx::Float64
    accy::Float64
end

function distance(p1::Particle, p2::Particle)::Float64
    return sqrt((p1.posx - p2.posx)^2 + (p1.posy - p2.posy)^2)
end

function update_position!(p::Particle)
    p.posx = p.posx + p.velx * DELTA_TIME;
    p.posy = p.posy + p.vely * DELTA_TIME; 
end

function update_velocity!(p::Particle)
    p.velx = p.velx + p.accx/p.mass * DELTA_TIME;
    p.vely = p.vely + p.accy/p.mass * DELTA_TIME; 
end

function update_acceleration!(p1::Particle, p2::Particle)
    dist = distance(p1, p2);
    f = CONST_G*(p1.mass * p2.mass)/(dist^2);

    p1.accx = p1.accx + f * (p2.posx - p1.posx) / dist 
    p1.accy = p1.accy + f * (p2.posy - p1.posy) / dist
end

function reset_force!(p::Particle)
    p.accx = 0.0
    p.accy = 0.0
end

function main()
    nb_particles = parse(Int64, ARGS[1])
    nb_timesteps = parse(Int64, ARGS[2])
    before = time()
    Random.seed!(0)
    particles = 1:nb_particles .|> (x -> Particle(rand(Float64, 1) * 100.0..., rand(Float64, 6)...))

    before_compute = time()
    for t in 1:nb_timesteps
        for i in 1:nb_particles
            accx = zeros(Float64, nthreads())
            accy = zeros(Float64, nthreads())
            @threads for j in 1:nb_particles
                if i != j
                    dist = distance(particles[i], particles[j]);
                    f = CONST_G*(particles[i].mass * particles[j].mass)/(dist^2);
                    accx[threadid()] += f * (particles[j].posx - particles[i].posx) / dist 
                    accy[threadid()] += f * (particles[j].posy - particles[i].posy) / dist
                end
            end
            particles[i].accx = sum(accx)
            particles[i].accy = sum(accy)
        end

        @threads for i in 1:nb_particles
            update_velocity!(particles[i])
            update_position!(particles[i])
            reset_force!(particles[i])
        end

    end
    fin = time()
    duration_ete = fin - before
    duration_compute = fin - before_compute
    println("$(duration_ete), $(duration_compute), $(particles[1].posx)")
end

main()

File: benchmark_scripts/nbody-par/py/nbody.py
import sys
import time
from random import random, seed
from math import sqrt, ceil
from multiprocessing.pool import ThreadPool

DELTA_TIME=1
CONST_G=1


class Particle:
    def __init__(self, mass, posx, posy, velx, vely, accx, accy):
        self.mass = mass
        self.posx = posx
        self.posy = posy
        self.velx = velx
        self.vely = vely
        self.accx = accx
        self.accy = accy

    def distance(self, other):
        return sqrt((self.posx-other.posx)**2 + (self.posy-other.posy)**2)

    def update_position(self):
        self.posx += self.velx * DELTA_TIME
        self.posy += self.vely * DELTA_TIME

    def update_velocity(self):
        self.velx += self.accx/self.mass * DELTA_TIME
        self.vely += self.accy/self.mass * DELTA_TIME

    def update_acceleration(self, other):
        dist = self.distance(other)
        f = CONST_G * (self.mass * other.mass) / (dist**2)
        self.accx += f * (other.posx - self.posx) / dist
        self.accy += f * (other.posy - self.posy) / dist

    def reset_force(self):
        self.accx = 0.0
        self.accy = 0.0

def random_particle():
    return Particle(100.0*random(),\
                    random(), random(),\
                    random(), random(),\
                    random(), random())

def worker_compute(tt):
    particles, i, start, end = tt
    nb_particles = len(particles)
    accx = 0.0
    accy = 0.0
    for j in range(start, min(end, nb_particles)):
        if i != j:
            dist = particles[i].distance(particles[j])
            f = CONST_G * (particles[i].mass * particles[j].mass) / (dist**2)
            accx += f * (particles[j].posx - particles[i].posx) / dist
            accy += f * (particles[j].posy - particles[i].posy) / dist
    return (accx, accy)

def worker_update(tt):
    particles, start, end = tt
    for i in range(start, min(end, len(particles))):
        particles[i].update_velocity()
        particles[i].update_position()
        particles[i].reset_force()

def main():
    args = sys.argv
    # assert len(args) == 4
    nb_particles = int(args[1])
    nb_timesteps = int(args[2])
    num_threads = int(args[3])
    start_data = time.time()
    w = ceil(1.0 * nb_particles / num_threads)

    seed(0)
    particles = [random_particle() for _ in range(nb_particles)]

    start_compute = time.time()
    with ThreadPool(processes=num_threads) as pool:
        for t in range(1,nb_timesteps+1):
            for i in range(nb_particles):
                results = pool.map(worker_compute, [(particles, i, start, start + w) for start in range(0, nb_particles,w)])
                acc = [sum(tup) for tup in zip(*results)]
                particles[i].accx = acc[0]
                particles[i].accy = acc[1]

            _ = pool.map(worker_update, [(particles, start, start + w) for start in range(0, nb_particles,w)])
    end = time.time()
    duration_ete = end - start_data
    duration_sort = end - start_compute
    print(f"{duration_ete},{duration_sort},{particles[0].posx}")

if __name__ == "__main__":
    main()
